components/admin_console/admin_settings.tsx

```diff
@@ -0,0 +1,258 @@
+import FormError from 'components/form_error';
+
+import AdminHeader from 'components/widgets/admin_console/admin_header';
+
+type Props = {
+    config?: object;
+    environmentConfig?: object;
+    setNavigationBlocked?: (blocked: boolean) => void;
+    updateConfig?: (config: object) => {data: object; error: ClientErrorPlaceholder};
+}
+
+type State = {
+    saveNeeded: boolean;
+    saving: boolean;
+    serverError: string|null;
+    serverErrorId?: string;
+    errorTooltip: boolean;
+}
+
+type StateKeys = keyof State;
+
+// Placeholder type until ClientError is exported from redux.
+// TODO: remove ClientErrorPlaceholder and change the return type of updateConfig
+type ClientErrorPlaceholder = {
+    message: string;
+    server_error_id: string;
+}
+
+export default abstract class AdminSettings extends React.Component<Props, State> {
+    public constructor(props: Props) {
+        super(props);
+        const stateInit = {
+            saveNeeded: false,
+            saving: false,
+            serverError: null,
+            errorTooltip: false,
+        };
+        if (props.config) {
+            this.state = Object.assign(this.getStateFromConfig(props.config), stateInit);
+        } else {
+            this.state = stateInit;
+        }
+    }
+
+    protected abstract getStateFromConfig(config: object): State;
+
+    protected abstract getConfigFromState(config: object): object;
+
+    protected abstract renderTitle(): React.ReactElement;
+
+    protected abstract renderSettings(): React.ReactElement;
+
+    protected handleSaved?: ((config: object) => React.ReactElement);
+
+    protected canSave?: () => boolean;
+
+    private closeTooltip = () => {
+        this.setState({errorTooltip: false});
+    }
+
+    private openTooltip = (e: React.MouseEvent) => {
+        const elm: HTMLElement|null = e.currentTarget.querySelector('.control-label');
+        if (elm) {
+            const isElipsis = elm.offsetWidth < elm.scrollWidth;
+            this.setState({errorTooltip: isElipsis});
+        }
+    }
+
+    private handleChange = (id: StateKeys, value: any) => {
+        this.setState((prevState) => ({
+            ...prevState,
+            saveNeeded: true,
+            [id]: value,
+        }));
+
+        if (this.props.setNavigationBlocked) {
+            this.props.setNavigationBlocked(true);
+        }
+    };
+
+    private handleSubmit = (e: React.SyntheticEvent) => {
+        e.preventDefault();
+
+        this.doSubmit();
+    }
+
+    private doSubmit = async (callback?: () => void) => {
+        this.setState({
+            saving: true,
+            serverError: null,
+        });
+
+        // clone config so that we aren't modifying data in the stores
+        let config = JSON.parse(JSON.stringify(this.props.config));
+        config = this.getConfigFromState(config);
+
+        if (this.props.updateConfig) {
+            const {data, error} = await this.props.updateConfig(config);
```

Comment:
On the original file, this function and several others were called without any kind of check on several non-required function props. Therefore, if these props were not given, Javascript would have crashed and thrown an exception, mostly silent to the user, but still an exception.
My approach so far was to silently go around the function if it is not given. Nevertheless, this might not be the expected behaviour. Specially on this part of the code, if we do not receive the updateConfig function, the configuration will not save, and several related state variables would not update neither.
What should be the approach for this case in particular, and other similar cases?